src/pilot/raid.patch

```diff
@@ -17,3 +45,51 @@
  
      max_volume_size_mb = _max_volume_size_mb(
          logical_disk['raid_level'], selected_disks, free_space_mb,
+@@ -624,7 +633,7 @@
+     return filtered_disks
+ 
+ 
+-def _commit_to_controllers(node, controllers):
++def _commit_to_controllers(node, controllers, realtime=False):
+     """Commit changes to RAID controllers on the node."""
+ 
+     if not controllers:
+@@ -641,10 +650,10 @@
+         # Do a reboot only for the last controller
+         if controller == controllers[-1]:
+             job_id = commit_config(node, raid_controller=controller,
+-                                   reboot=True)
++                                   reboot=True, realtime=realtime)
+         else:
+             job_id = commit_config(node, raid_controller=controller,
+-                                   reboot=False)
++                                   reboot=False, realtime=realtime)
+ 
+         LOG.info('Change has been committed to RAID controller '
+                  '%(controller)s on node %(node)s. '
+@@ -729,10 +738,12 @@
+         logical_disks_to_create = _filter_logical_disks(
+             logical_disks, create_root_volume, create_nonroot_volumes)
+ 
++        realtime=True
++        reboot_is_required = drac_constants.RebootRequired.true
+         controllers = set()
+         for logical_disk in logical_disks_to_create:
+             controllers.add(logical_disk['controller'])
+-            create_virtual_disk(
++            result = create_virtual_disk(
+                 node,
+                 raid_controller=logical_disk['controller'],
+                 physical_disks=logical_disk['physical_disks'],
+@@ -742,7 +753,10 @@
+                 span_length=logical_disk.get('span_length'),
+                 span_depth=logical_disk.get('span_depth'))
+ 
+-        return _commit_to_controllers(node, list(controllers))
++            if result['is_reboot_required'] == reboot_is_required:
++                realtime = False
++
++        return _commit_to_controllers(node, list(controllers), realtime)
```

Comment:
If I understand correctly, this would not correctly work when a node has multiple RAID controllers with differing real time capabilities. For example:

* Two (2) RAID controllers
* One has real time, so it returns `result['is_reboot_required']` set to `drac_constants.RebootRequired.optional`
* One does not have real time, so it returns `result['is_reboot_required']` set to `drac_constants.RebootRequired.true`

At the end of the loop, the variable `realtime` will be set to `False`, and `_commit_to_controllers()` will not use real time at all. The RAID controller that has real time time would still encounter the bug this PR is attempting to fix.

Reply:
You are making huge assumptions on how the iDRAC will behave when a server has multiple RAID controllers, some of which support realtime mode and some of which don't.  Before we start writing code to handle that scenario, we would need to test the behavior of hardware in that configuration.  I'm not even sure that such a configuration is valid.
Also, JetPack doesn't support having multiple RAID controllers in a server, so your point is moot.

Reply:
A Dell EMC RAID and WS-Man subject matter expert confirmed that an iDRAC can manage multiple RAID controllers, some of which support real-time mode and some of which don't. And, as the ironic iDRAC driver code in question implements, virtual volumes can be created on them at the same time.

An example is a server that has two (2) RAID controllers, a PERC and BOSS. Recent generations of PERC support real-time, however, BOSS does not. Real-time and reboot should be specified when creating the two (2) configuration jobs, respectively. The iDRAC executes the real-time job for the PERC first. After that job completes, it reboots the system and executes the staged job for the BOSS.

I understand that configuration may not be supported by JetPack, in which case, this downstream patch is fine, as is. However, please note this change should not be pushed to upstream ironic without testing against such a configuration.